File: FsharpTodoApp.Infrastructure/Persistence/DataModels/DataModelBase.fs
namespace FsharpTodoApp.Infrastructure.Persistence.DataModels

open FsharpTodoApp.Domain.Common.ValueObjects
open FsharpTodoApp.Domain.Common.Entities
open FsharpTodoApp.Persistence.DataModels

module DataModelBase =
    let dehydrateDeletionState (dataModel: DataModelBase) (deletedAudit: DeletedAudit) =
        DeletedAudit.value deletedAudit
        |> Option.iter (fun audit ->
            dataModel.DeletedAt <- Some audit.Timestamp |> Option.toNullable
            dataModel.DeletedBy <- Some audit.UserInfo.UserName |> Option.defaultValue null)

    let dehydrate (dataModel: DataModelBase) (entityBase: EntityBase) =
        dataModel.Id <- entityBase.IdSet.DbId
        dataModel.PublicId <- entityBase.IdSet.PublicId

        CreatedAudit.value entityBase.CreatedAudit
        |> fun audit ->
            dataModel.CreatedAt <- audit.Timestamp
            dataModel.CreatedBy <- audit.UserInfo.UserName

        UpdatedAudit.value entityBase.UpdatedAudit
        |> Option.iter (fun audit ->
            dataModel.UpdatedAt <- Some audit.Timestamp |> Option.toNullable
            dataModel.UpdatedBy <- Some audit.UserInfo.UserName |> Option.defaultValue null)

        entityBase.DeletedAudit |> dehydrateDeletionState dataModel

File: FsharpTodoApp.Persistence/DataModels/DataModelBase.cs
using Microsoft.EntityFrameworkCore;

namespace FsharpTodoApp.Persistence.DataModels;

public abstract class DataModelBase
{
    public long Id { get; set; }
    public Guid PublicId { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public DateTime? DeletedAt { get; set; }

    public string CreatedBy { get; set; } = string.Empty;
    public string? UpdatedBy { get; set; }
    public string? DeletedBy { get; set; }


    public static void OnModelCreating<T>(ModelBuilder modelBuilder, string tableName)
        where T : DataModelBase
    {
        modelBuilder.Entity<T>().HasKey(e => e.Id);

        modelBuilder.Entity<T>().HasIndex(e => e.PublicId).IsUnique();

        modelBuilder.Entity<T>().HasQueryFilter(e => e.DeletedAt == null);

        modelBuilder.Entity<T>().ToTable(tableName);
    }
}
